esmvalcore/preprocessor/_multimodel.py

```diff
@@ -248,115 +98,190 @@ def _unify_time_coordinates(cubes):
         cube.coord('time').guess_bounds()
 
 
-def _get_time_slice(cubes, time):
-    """Fill time slice array with cubes' data if time in cube, else mask."""
-    time_slice = []
-    for cube in cubes:
-        cube_time = cube.coord('time').points
-        if time in cube_time:
-            idx = int(np.argwhere(cube_time == time))
-            subset = cube.data[idx]
-        else:
-            subset = np.ma.empty(list(cube.shape[1:]))
-            subset.mask = True
-        time_slice.append(subset)
-    return time_slice
+def _time_coords_are_aligned(cubes):
+    """Return `True` if time coords are aligned."""
+    first_time_array = cubes[0].coord('time').points
+
+    for cube in cubes[1:]:
+        other_time_array = cube.coord('time').points
+        if not np.array_equal(first_time_array, other_time_array):
+            return False
+
+    return True
+
+
+def _subset(cube, time_points):
+    """Subset cube to given time range."""
+    begin = cube.coord('time').units.num2date(time_points[0])
+    end = cube.coord('time').units.num2date(time_points[-1])
+    constraint = iris.Constraint(time=lambda cell: begin <= cell.point <= end)
+    return cube.extract(constraint)
+
+
+def _extend(cube, time_points):
+    """Extend cube to a specified time range.
+
+    If time points are missing before the start/after the end of the
+    time range, cubes for each missing time pointwith masked data will
+    be added to pad the time range to match `time_points`. This method
+    supports lazy operation.
+    """
+    cube.coord('time').bounds = None
+    cube_points = cube.coord('time').points
+
+    begin = cube_points[0]
+    end = cube_points[-1]
+
+    pad_begin = time_points[time_points < begin]
+    pad_end = time_points[time_points > end]
+
+    if (len(pad_begin)) == 0 and (len(pad_end) == 0):
+        return cube
+
+    template_cube = cube[:1].copy()
```

Comment:
Good question! We want to get a cube with the time axis intact. `cube[0]` returns a scalar cube.

Reply:
Just tested this to make sure that `.copy` is also a lazy operation if the input cube is lazy.

Reply:
OK about indexing, I thought you were selecting the 0th time cube, if you do `cube[0,]` or `cube[0]` that shouldn't return a scalar cube, but the 0th time cube of the original cube - yes, the time axis would be scalar or better said, vanish so if you still need it yes, do the [:1] indexing :+1: About copy - yes, a copy doesn't realize the data but even if unrealized, it still can be a hefty cube in memory so that's why I am on about trying not to make too many copies

Reply:
the problem is that weighted collapse is still not lazy in iris AFASK